tutorials/data-science-preprocessing/index.md

```diff
@@ -0,0 +1,437 @@
+
+[embedmd]:# (src/preprocessing/clean.py /def main/ / p\.run.*/)
+```py
+def main(src_path, dest_table, pipeline_args):
+    p = apache_beam.Pipeline(argv=pipeline_args)
+
+    value = p | 'Read JSON' >> apache_beam.Read(JsonFileSource(src_path))
+
+    value |= (
+        'Remove records that lack location or year data' >>
+        apache_beam.FlatMap(discard_incomplete))
+
+    value |= (
+        'Convert string values to their types' >>
+        apache_beam.Map(convert_types))
+
+    value |= (
+        'Filter bad data' >>
+        apache_beam.FlatMap(filter_suspicious))
+
+    value |= (
+        'Massage fields with "rec" prefix' >>
+        apache_beam.Map(massage_rec))
+
+    value |= (
+        'Dump data to BigQuery' >>
+        apache_beam.Write(apache_beam.io.BigQuerySink(
+            dest_table,
+            schema=', '.join([
+                'fall:STRING',
+                'year:INTEGER',
+                'nametype:STRING',
+                'mass:FLOAT',
+                'name:STRING',
+                'class:STRING',
+                'latitude:FLOAT',
+                'longitude:FLOAT',
+                'id:STRING',
+            ]),
+            create_disposition=(
+                apache_beam.io.BigQueryDisposition.CREATE_IF_NEEDED),
+            write_disposition=(
+                apache_beam.io.BigQueryDisposition.WRITE_TRUNCATE))))
+
+    p.run()
+```
+
+Note that this example takes advantage of the built-in [`BigQuerySink`][bq-sink]
+to output the result into BigQuery, to analyze later.
+
+[bq-sink]: https://github.com/apache/incubator-beam/tree/python-sdk/sdks/python#bigquery
+
+## Run the data cleansing pipeline
+
+A great thing about Apache Beam and Cloud Dataflow is that you can run it
+locally for sample datasets, or datasets that are otherwise relatively small.
+But then you can take the same code, add a couple extra flags, and run it using
+Cloud Dataflow which can handle orders of magnitude more data at a time.
+
+Because this sample uses a `BigQuerySink` for its output, you must first perform
+some setup:
+
+* First, make sure you're using the correct project:
+
+      $ gcloud config set project your-project-id
+      Updated property [core/project]
+      $
+
+* Now, create a BigQuery dataset using the `bq` command provided by the Google
+  Cloud SDK:
+
+      $ bq mk meteor_dataset
+      Dataset 'your-project-id:meteor_dataset' successfully created
+      $
+
+* To ensure that this example's dependencies don't conflict with any existing
+  libraries on your system, create a virtualenv and install the script
+  dependencies:
+
+      $ virtualenv venv
+      ...
+      $ source venv/bin/activate
+      (venv) $ pip install -r requirements.txt
+      ...
+      (venv) $
+
+With the BigQuery dataset created and your requirements installed, you are ready
+to run your pipeline.
+
+### Running locally
+
+    python clean.py meteors.json your-project-id:meteor_dataset.cleansed
```

Comment:
I suggest a quick introductory sentence here, both to make absolutely clear what's happening and to create a parallel structure for this section and the next one.